drivers/adc/adc_mcux_lpadc.c

```diff
@@ -57,13 +58,70 @@ struct mcux_lpadc_data {
 	lpadc_conv_command_config_t cmd_config[CONFIG_LPADC_CHANNEL_COUNT];
 };
 
+static int mcux_lpadc_acquisition_time_setup(const struct device *dev, uint16_t acq_time,
+					     lpadc_conv_command_config_t *cmd)
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	uint32_t adc_freq_hz = 0;
+	uint32_t conversion_factor = 0;
+	uint32_t acquisition_time_value = ADC_ACQ_TIME_VALUE(acq_time);
+	uint8_t acquisition_time_unit = ADC_ACQ_TIME_UNIT(acq_time);
+
+	if (ADC_ACQ_TIME_DEFAULT == acquisition_time_value) {
+		return 0;
+	}
+
+	/* If the acquisition time is expressed in ADC ticks, then directly compare
+	 * the acquisition time with configuration items (3, 5, 7, etc. ADC ticks)
+	 * supported by the LPADC. The conversion factor is set to 1 (means do not need
+	 * to convert configuration items from ADC ticks to nanoseconds).
+	 * If the acquisition time is expressed in microseconds or nanoseconds, First
+	 * calculate the ADC cycle based on the ADC clock, then convert the configuration
+	 * items supported by LPADC into nanoseconds, and finally compare the acquisition
+	 * time with configuration items. The conversion factor is equal to the ADC cycle
+	 * (means convert configuration items from ADC ticks to nanoseconds).
+	 */
+	if (ADC_ACQ_TIME_TICKS == acquisition_time_unit) {
+		conversion_factor = 1;
+	} else {
+		if (clock_control_get_rate(config->clock_dev, config->clock_subsys, &adc_freq_hz)) {
+			LOG_ERR("Get clock rate failed");
+			return -EINVAL;
+		}
+
+		conversion_factor = (uint32_t)((float)1000000000.0 / adc_freq_hz);
+
+		if (ADC_ACQ_TIME_MICROSECONDS == acquisition_time_unit) {
+			acquisition_time_value *= 1000;
+		}
+	}
+
+	if ((3 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK3;
+	} else if ((5 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK5;
+	} else if ((7 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK7;
+	} else if ((11 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK11;
+	} else if ((19 * conversion_factor) >= acquisition_time_value) {
+		cmd->sampleTimeMode = kLPADC_SampleTimeADCK19;
+	} else if ((35 * conversion_factor) >= acquisition_time_value) {
```

Comment:
this code is fine now, I much prefer this approach, but one thing I remember you said is you didn't know why to just pick the greatest option rather than the closest.... well, I don't know the reason either, maybe @anangl can clarify, but I think if you want to pick the closest value, that seems fine to me.... the way you can do it, is to check if the value is greater than the midpoint between the points, rather than the points itself, like this:

```
if ((4* conversion_factor) >= time) {
mode = 3
} else if (6*fact >= time) {
mode = 5
} else if(9 *factor >= time {
mode = 7
} ... etc
```
see what I mean about how it will pick then the closest value ?

Reply:
the way you provided to find the most closest options is also ok, but the time complexity of binary search is better than the ordered search. Of course, the amount of options is not large, so the performance of the two algorithms is similar.

And I guess @anangl may think we should ensure sufficient sampling time of the ADC, because the sampling time will affect the accuracy of the ADC, although this will result in a drop in sampling rate.